### PR TITLE
Energy Meter Improvements

### PR DESCRIPTION
**Pull request by tasodorff**
Extended energy meters and refactored some underlying abstract types.

Only one existing energy meter was impacted, with changes being to inherit fields from abstract concepts instead of through direct definition of fields on that type (EM_ION).

### CODE REVIEW

**Review by berkoben — CHANGES_REQUESTED** [review]

File: ontology/yaml/resources/METERS/entity_types/POWER_METER.yaml [hunk 1]
```diff
@@ -12,32 +12,69 @@
...
+EM_PWM_AVCM_PVCM:
+  description: "Power meter with phase level voltage and current monitoring (and averaging)."
+  is_canonical: true
+  implements:
+  - EM
+  - PWM
+  - AVCM
+  - PVCM
+
+EM_ION: # TODO: should rename to standard convention.
```
[the remark this answers is not in the document]
Comment (tasodorff) [reply ontology/yaml/resources/METERS/entity_types/POWER_METER.yaml:39]: fixed in new commit

File: ontology/yaml/resources/METERS/entity_types/POWER_METER.yaml [hunk 2]
```diff
@@ -12,32 +12,69 @@
...
+  - AVCM
+  - PLPM
+  - PVCM
+
+
+#############################
+### Non-Canonical Devices ###
+#############################
+
+EM_ONEOFF_1:
```
[the remark this answers is not in the document]
Comment (tasodorff) [reply ontology/yaml/resources/METERS/entity_types/POWER_METER.yaml:55]: i will remove them, they definitely are not valid canonical configurations.

**Review by berkoben — CHANGES_REQUESTED** [review]

File: ontology/yaml/resources/METERS/entity_types/ABSTRACT.yaml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  - line_frequency_sensor
+  uses:
+  - power_sensor
+  - energy_accumulator
+
+AVCM:
+  description: "Average phase (line) level voltage and current monitoring. "
+  is_abstract: true
+  uses:
+  - average_line_voltage_sensor # Line to line
```
[the remark this answers is not in the document]
Comment (tasodorff) [reply ontology/yaml/resources/METERS/entity_types/ABSTRACT.yaml:34]: i think that is fine. i will update.

[the remark this answers is not in the document]
Comment (berkoben) [reply ontology/yaml/resources/METERS/entity_types/ABSTRACT.yaml:34]: yes, and I see you've updated it.

**Review by berkoben — APPROVED** [review]

File: ontology/yaml/resources/METERS/entity_types/ABSTRACT.yaml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+### ABSTRACT TYPES
+
+
+PWM:
+  description: "Very basic power and enrgy monitoring only."
+  is_abstract: true
```
[the remark this answers is not in the document]
Comment (tasodorff) [reply ontology/yaml/resources/METERS/entity_types/ABSTRACT.yaml:20]: There is probably a difference between the way that you and I think about the function of 'canonical': I don't consider something canonical unless it represents a concrete device (such as a VAV_SD, which we will find in the wild); there may be wonky devices where we may argue over if its truly canonical, but they are all inevitably concrete. An abstract type, by comparison, does not have any concrete manifestation in and of itself -- it is always an abstraction of function (or something else, such as a convenient grouping of fields) and must therefore always be implemented by a concrete canonical type, even if it is a 1:1 mapping.

**Pull request merged**